### PR TITLE
add option for json decoder to return unwrapped

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

**Comment by id** [comment]
Bump version?

File: src/avro_json_decoder.erl [hunk 1]
```diff
@@ -47,37 +49,44 @@ decode_schema(Json) ->
...
-decode_value(JsonValue, Schema, ExtractTypeFun) ->
-  parse_value(mochijson3:decode(JsonValue), Schema, ExtractTypeFun).
-
-%% Experimental support for jsonx decoding.
 %% jsonx is about 12 times faster than mochijson3 and almost compatible
 %% with it. The one discovered incompatibility is that jsonx performs more
 %% strict checks on incoming json strings and can fail on some cases
 %% (for example, on non-ascii symbols) while mochijson3 can parse such strings
 %% without problems.
 %% Current strategy is to use jsonx as the main parser and fall back to
```
Comment (id) [inline src/avro_json_decoder.erl]: Doesn't look like jsonx is the main parser, one should explicitly specify it in Options.
May be use `{decoder, jsonx}` in decode_value/3?

File: src/avro_json_decoder.erl [hunk 2]
```diff
@@ -47,37 +49,44 @@ decode_schema(Json) ->
...
+          {error, _Err, _Pos} ->
+            mochijson3:decode(JsonValue);
+          Decoded ->
+            Decoded
+        end;
+      _ ->
+        mochijson3:decode(JsonValue)
+    end,
+  IsWrapped = case lists:keyfind(is_wrapped, Options) of
+                {is_wrapped, V} -> V;
```
Comment (id) [inline src/avro_json_decoder.erl]: Need more docs explaining what does `is_wrapped` mean.

File: src/avro_json_decoder.erl [hunk 3]
```diff
@@ -266,68 +275,70 @@ type_from_name(_)                 -> undefined.
...
+  parse(Value, Type, ExtractFun, _IsWrapped = true).
 
-parse_value(V, Type, _ExtractFun) when ?AVRO_IS_BOOLEAN_TYPE(Type) andalso
-                                       is_boolean(V) ->
+parse(Value, TypeName, ExtractFun, IsWrapped) when is_list(TypeName) ->
+  %% Type is defined by its name
+  Type = ExtractFun(TypeName),
+  parse(Value, Type, ExtractFun, IsWrapped);
+parse(Value, Type, _ExtractFun, IsWrapped) when ?AVRO_IS_PRIMITIVE_TYPE(Type) ->
+  WrappedValue = parse_prim(Value, Type),
```
Comment (id) [inline src/avro_json_decoder.erl]: So here we wrap first and then unwrap when IsWrapped =:= false.

What if parse_prim would return an unwrapped value + Fun to wrap it? Then call Fun when IsWrapped =:= true.

Comment (zmstone) [inline src/avro_json_decoder.erl]: wrapping / striping primitives are quite cheap.
don't think it worth the effort to add function clauses for them.

Comment (id) [inline src/avro_json_decoder.erl]: No, not adding clauses. What I mean is everything is the same except parse_prim clauses return 
e.g. {fun avro_primitive:int/1, V}, and here you call WrapFun(V) when IsWrapped =:= true.

Minor thing, agreed.

**Comment by id** [comment]
👍

**Comment by zmstone** [comment]
@id good to  merge ?

**Comment by id** [comment]
Yes, go ahead.

**Pull request merged**